src/lib/ppx_register_event/register_event.ml

```diff
@@ -0,0 +1,201 @@
+open Core_kernel
+open Ppxlib
+
+let deriver = "register_event"
+
+(* module in Structured_events library *)
+let module_name = "Structured_events"
+
+let hash name =
+  let open Digestif.SHA256 in
+  let ctx0 = init () in
+  let ctx1 = feed_string ctx0 name in
+  get ctx1 |> to_raw_string
+
+(* we omit an explicit `Core_kernel`, because the Ppx_deriving_yojson functions
+   generate uses of `Result` without it, so that module will need to be opened in
+   any case
+*)
+let core_kernel_result f = String.concat ["Result"; f] ~sep:"."
+
+let type_ext_str ~options ~path (ty_ext : type_extension) =
+  let add_module_qualifier s = String.concat ~sep:"." [module_name; s] in
+  let ctor, ctor_args, label_decls =
+    match ty_ext.ptyext_constructors with
+    (* record argument *)
+    | [ { pext_name
+        ; pext_kind= Pext_decl ((Pcstr_record labels as record), None)
+        ; _ } ] ->
+        (pext_name.txt, record, labels)
+    (* no arguments *)
+    | [{pext_name; pext_kind= Pext_decl ((Pcstr_tuple [] as tuple), None); _}]
+      ->
+        (pext_name.txt, tuple, [])
+    | _ ->
+        Location.raise_errorf ~loc:ty_ext.ptyext_path.loc
+          "Constructor in type extension must take a single record argument, \
+           or no argument"
+  in
+  let label_names =
+    List.map label_decls ~f:(fun {pld_name= {txt; _}; _} -> txt)
+  in
+  let has_record_arg = not @@ List.is_empty label_names in
+  let msg, msg_loc =
```

Comment:
Don't need to do this now, but it would be nice if we used the expression passed as `msg` directly. For example

```ocaml
module F (module X : struct
    type t
    val modname : string
end) = struct
  type  Structured_events.t += Error {data: X.t}
    [@@deriving register_event {msg= "F("^X.modname^").Error: Could not do thing, had $data"}]
end
```

Reply:
Done!